src/blockdata/script.rs

```diff
@@ -643,107 +678,339 @@ impl Script {
+    /// script tree merkle root.
+    pub fn new_v1_p2tr<C: Verification>(secp: &Secp256k1<C>, internal_key: UntweakedPublicKey, merkle_root: Option<TapBranchHash>) -> Self {
+        let (output_key, _) = internal_key.tap_tweak(secp, merkle_root);
+        ScriptBuf::new_witness_program(WitnessVersion::V1, &output_key.serialize())
+    }
+
+    /// Generates P2TR for key spending path for a known [`TweakedPublicKey`].
+    pub fn new_v1_p2tr_tweaked(output_key: TweakedPublicKey) -> Self {
+        ScriptBuf::new_witness_program(WitnessVersion::V1, &output_key.serialize())
+    }
+
+    /// Generates P2WSH-type of scriptPubkey with a given hash of the redeem script.
+    pub fn new_witness_program(version: WitnessVersion, program: &[u8]) -> Self {
+        Builder::new()
+            .push_opcode(version.into())
+            .push_slice(program)
+            .into_script()
+    }
+
+    /// Generates OP_RETURN-type of scriptPubkey for the given data.
+    pub fn new_op_return(data: &[u8]) -> Self {
+        Builder::new()
+            .push_opcode(opcodes::all::OP_RETURN)
+            .push_slice(data)
+            .into_script()
+    }
+
+    /// Converts the script into a byte vector.
+    pub fn into_bytes(self) -> Vec<u8> { self.0 }
+
+    /// Computes the P2SH output corresponding to this redeem script.
+    pub fn to_p2sh(&self) -> ScriptBuf {
+        ScriptBuf::new_p2sh(&self.script_hash())
+    }
+
+    /// Returns the script code used for spending a P2WPKH output if this script is a script pubkey
+    /// for a P2WPKH output. The `scriptCode` is described in [BIP143].
+    ///
+    /// [BIP143]: <https://github.com/bitcoin/bips/blob/99701f68a88ce33b2d0838eb84e115cef505b4c2/bip-0143.mediawiki>
+    pub fn p2wpkh_script_code(&self) -> Option<ScriptBuf> {
+        if !self.is_v0_p2wpkh() {
+            return None
+        }
+        let script = Builder::new()
+            .push_opcode(opcodes::all::OP_DUP)
+            .push_opcode(opcodes::all::OP_HASH160)
+            .push_slice(&self.as_bytes()[2..]) // The `self` script is 0x00, 0x14, <pubkey_hash>
+            .push_opcode(opcodes::all::OP_EQUALVERIFY)
+            .push_opcode(opcodes::all::OP_CHECKSIG)
+            .into_script();
+
+        Some(script)
     }
 
     /// Creates an assembly decoding of the script in the given byte slice.
     pub fn bytes_to_asm(script: &[u8]) -> String {
         let mut buf = String::new();
-        Script::bytes_to_asm_fmt(script, &mut buf).unwrap();
+        bytes_to_asm_fmt(script, &mut buf).unwrap();
         buf
     }
 
-    /// Returns the assembly decoding of the script.
-    pub fn asm(&self) -> String {
-        Script::bytes_to_asm(self.as_ref())
+    /// Adds a single opcode to the script.
+    pub fn push_opcode(&mut self, data: opcodes::All) {
+        self.0.push(data.to_u8());
+    }
+
+    /// Adds instructions to push some arbitrary data onto the stack.
+    pub fn push_slice(&mut self, data: &[u8]) {
```

Comment:
Its not clear to me why these methods are on `ScriptBuf` when we have the `Builder`, can you explain the use case please?

Reply:
Yes, wanted to document better, but forgot. `Builder` has more ergonomic API for build-and-hand-over code but it requires `self`, where `ScriptBuf` takes `&mut self` instead.

Reply:
Cool, I did not notice that detail. Thanks.